### PR TITLE
Adjust osu! autoplay generation to ensure constant SPM with rate-changing mods

### LINKED ISSUE #10283: Autoplay spinning speed depends on rate-changing mods
**Describe the bug:** 
The bonus score of a spinner with DT mod is capped by the maximum attainable spinner bonus score (i.e. at 477 rpm / auto). Notice in the video when the map is played with DT, the first spinner does not go pass 850 for its bonus score even with auto mod, which is the spinner score auto mod got for that spinner when played with NM

**Screenshots or videos showing encountered issue:** 
https://streamable.com/m3l50j

**osu!lazer version:** 
2020.925.0

**Logs:**
[database.log](https://github.com/ppy/osu/files/5295003/database.log)
[performance-audio.log](https://github.com/ppy/osu/files/5295005/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/5295006/performance-draw.log)
[performance-update.log](https://github.com/ppy/osu/files/5295007/performance-update.log)
[performance.log](https://github.com/ppy/osu/files/5295008/performance.log)
[runtime.log](https://github.com/ppy/osu/files/5295009/runtime.log)

### PR DESCRIPTION
**Pull request by bdach**
Closes #10283.

A few notes to explain things:

The first thing that this PR does to achieve its end result is allow autoplay implementations to access the list of mods used to generate the autoplay replay for. I've gone ahead and marked the old `CreateReplayScore()` method as obsolete in favour of the one that accepts mods - consumers can always ignore the mods. Curious if you agree with this choice, and I'll split this part of the PR (essentially the first commit) on request.

Secondly, most of the pieces in place to achieve this (the same way that stable seems to do) were in place already in `OsuAutoGeneratorBase`, in the way of two methods: `ApplyModsToRate(double)` and `ApplyModsToTime(double)`. Unfortunately, `ModTimeRamp` - which was not in stable - is a major complication to these methods, as the track rate in it depends on track time, which kills the simple calculations.

I've managed to somewhat salvage that structure by moving stuff around. All call sites of `ApplyModsToRate()` now need to accept a track time instant to be able to transform time correctly - in particular some things that were constants previously, now can't be for exactly that reason, and are usually replaced with private/protected methods that calculate these constants for a given time instant (see `FrameDelay` and `reactionTime`).

The worse part was `ApplyModsToTime()`. The transformation of track time to real time for `ModTimeRamp` is - I'm pretty sure - non-linear. [I thought I had the formula for it right](https://github.com/ppy/osu/issues/10283#issuecomment-771241482), but it was just slightly off, and it was inscrutable anyway. Thankfully today I realised there is an alternative - because all usages of that method were usually called on time *deltas*, not actual time instants, then I can perform a local approximation for that delta by just dividing by the track rate at a given time instant. It's not perfectly correct, but seems to work well enough in practice.

I've both tested this manually [on this map](https://osu.ppy.sh/beatmapsets/679738) and added automated test cases. Note that scrubbing the replay will break this because it's presumably playing back past frames with a future rate, but that's a separate issue and this pull is too large as-is to meaningfully address that anyway.

### CODE REVIEW

**Comment by peppy** [comment]
I think I can get behind the general structure of these changes (definitely better than the alternatives we've tried previously), but can you elaborate on the seeking hypothesis you have? Is the "applying a future rate" something happening elsewhere as a completely separate issue? Just trying to imagine what is going on there so I can ignore the issue while reviewing this if feasible.

**Comment by bdach** [comment]
I'd have to take a proper second look to confirm what might be going on for sure, but my working surface level hypothesis was that `ModTimeRamp` instantaneously applies track rate after seeking, therefore desyncing the replay with the actual gameplay (frames in the past relative to the seek point now get the future rate multiplier valid at the seek point itself - see `RateAdjustedRotation`.

**Comment by peppy** [comment]
Aha, I think I can understand what's going on there, with the gameplay clock vs the audio clock.

File: osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs [hunk 1]
```diff
@@ -62,7 +62,8 @@ public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset
             inputManager.AllowUserCursorMovement = false;
 
             // Generate the replay frames the cursor should follow
-            replayFrames = new OsuAutoGenerator(drawableRuleset.Beatmap).Generate().Frames.Cast<OsuReplayFrame>().ToList();
+            var mods = (IReadOnlyList<Mod>)drawableRuleset.Dependencies.Get(typeof(IReadOnlyList<Mod>));
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:65]: wondering if we should expose this on `DrawableRuleset` instead... @smoogipoo thoughts?

File: osu.Game/Rulesets/Mods/ModTimeRamp.cs [hunk 2]
```diff
@@ -75,17 +75,24 @@ public virtual void ApplyToBeatmap(IBeatmap beatmap)
             finalRateTime = firstObjectStart + FINAL_RATE_PROGRESS * (lastObjectEnd - firstObjectStart);
         }
 
+        public double ApplyToRate(double time, double rate = 1)
+        {
+            double amount = (time - beginRampTime) / Math.Max(1, finalRateTime - beginRampTime);
+            double ramp = InitialRate.Value + (FinalRate.Value - InitialRate.Value) * Math.Clamp(amount, 0, 1);
+
+            // round the end result to match the bindable SpeedChange's precision, in case this is called externally.
+            return rate * Math.Round(ramp, 2);
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModTimeRamp.cs:84]: I think this is going to change the runtime behaviour of the mod to be less smooth in the ramping, no? Not sure if it's enough to be audible (and whether it sounds bad) yet, but worth noting.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:65]: Agree with exposing from `DrawableRuleset`.

File: osu.Game/Rulesets/Mods/ModTimeRamp.cs (hunk 2 above)
Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:84]: I think that was the case already. The bindable was rounding in its own anyways due to the precision set.

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:84]: Oh. Well then. I plan on testing this out to see how it feels sounds, but should be fine in that case.

**Review by peppy — APPROVED** [review]

**Pull request merged**